### ISSUE TITLE
[YSQL] TPCC tests with restore failing with "Query error: schema version mismatch for table 000033e800003000800000000000401b: expected 1, got 0"

### ISSUE THREAD

**Issue by sonalsagarwal** [yugabyte/yugabyte-db#13168]
Jira Link: [DB-2853](https://yugabyte.atlassian.net/browse/DB-2853)
### Description

The TPCC perf regression for 1k WH on 2.15.1.0-b122 is failing in execute phase with :
`"Query error: schema version mismatch for table 000033e800003000800000000000401b: expected 1, got 0"`
Stack trace :
```
java.lang.RuntimeException: Unexpected fatal, error in 'Worker<9440>' when executing 'StockLevel/05'
    at com.oltpbenchmark.api.Worker.doWork(Worker.java:584)
    at com.oltpbenchmark.api.Worker.run(Worker.java:368)
    at java.lang.Thread.run(Thread.java:750)
Caused by: org.postgresql.util.PSQLException: ERROR: Query error: schema version mismatch for table 000033e800003000800000000000401b: expected 1, got 0
    at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2468)
    at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2211)
    at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:309)
    at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:446)
    at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:370)
    at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:149)
    at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:108)
    at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52)
    at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java)
    at com.oltpbenchmark.jdbc.InstrumentedPreparedStatement.executeQuery(InstrumentedPreparedStatement.java:35)
    at com.oltpbenchmark.benchmarks.tpcc.procedures.StockLevel.run(StockLevel.java:98)
    at com.oltpbenchmark.api.Worker.executeWork(Worker.java:609)
    at com.oltpbenchmark.api.Worker.doWork(Worker.java:506)
    ... 2 more
```
The data is loaded using restore from s3.

The tests has started failing from 2.15.1.0-b122. The test is running successfully on 2.15.1.0-b121
The suspected check-in : https://phabricator.dev.yugabyte.com/rYBDB61985b6e0cf2ebbcab943c66ddda8138c0075dc1
 phabricator diff : https://phabricator.dev.yugabyte.com/D17276

**Comment by arybochkin** [comment]
More details:
1) Restoration is happening in three main steps:
a) DDL is restored from YSQLDump
b) Tables structure is adjusted in accordance with backuped tablets structure (partitioning, columns, etc)
c) Data is restored
2) Schema can be changed on the step 1.a currently for two cases only:
a) Columns ids was changed and should be restored, [code link](https://github.com/yugabyte/yugabyte-db/blob/4e74c5991925f2604e11f6c49127cd4ad6dfc7b3/ent/src/yb/master/catalog_manager_ent.cc#L1808)
b) Partitioning version was changed and the original one should be restored, [code link](https://github.com/yugabyte/yugabyte-db/blob/4e74c5991925f2604e11f6c49127cd4ad6dfc7b3/ent/src/yb/master/catalog_manager_ent.cc#L1843)
3) Looks like the test is rather straightforward and the case 2.a is never hit, but actually we will face exactly the same error if that will happen
4) The case 2.b is hit because partitioning version was changed with this commit globally; above are the steps how to reproduce for every build / commit starting from 2.15.1.0-b122 / 61985b6e0cf2ebbcab943c66ddda8138c0075dc1
5) PG detects the schema version has been changed, but due to the current design does not invalidates table entry in the cache when this's happening in a transactional block, [code link](https://github.com/yugabyte/yugabyte-db/blob/4e74c5991925f2604e11f6c49127cd4ad6dfc7b3/src/postgres/src/backend/tcop/postgres.c#L3838)
6) For the non-transactional queries PG invalidates table entry in the cache and re-sends the request by itself, [code link](https://github.com/yugabyte/yugabyte-db/blob/4e74c5991925f2604e11f6c49127cd4ad6dfc7b3/src/postgres/src/backend/tcop/postgres.c#L3874), this can be confirmed with a log (added more traces, starting with `AR-`), at least message about invalidation will be observed:
```
[ts-1] I0707 08:33:31.971633 24253 tablet.cc:1507] AR-HandlePgsqlReadRequest:
       table = 000033e5000030008000000000004100,
       table_info::schema_version = 1, read_requst::schema_version = 0
[ts-1] E0707 08:33:31.971715 24253 tablet.cc:1518] AR-HandlePgsqlReadRequest:
       table = 000033e5000030008000000000004100 schema version mismatch! expected = 1, got = 0
[ts-1] 2022-07-07 08:33:31.976 UTC [24497] LOG:  
       invalidating table cache entry 000033e5000030008000000000004100
[ts-1] I0707 08:33:31.996312 24252 tablet.cc:1507] AR-HandlePgsqlReadRequest:
       table = 000033e5000030008000000000004100,
       table_info::schema_version = 1, read_requst::schema_version = 1
```

Having all of this, what is observed in the test is the expected behaviour in accordance with current design.
Possible solutions:
1) Do not update schema version for partitioning_version change during restore -- not OK, possible may lead to a situation that other masters/tservers won't get up-to-date info in time; also this does not solve the case for changing columns ids during restoration. By the way, two schema's having difference only in partitioning_version are considered equal (partitioning_version is ignored during comparison).
2) Force restore procedure to make PG invalidate it's cache. Looks like it should be done for all connections (having several ybsqlsh opened, a non-transactional select or reconnect is required for all shells), but not 100% sure
3) After restore is done a simple non-transactional select can be done to force to update table entry in the cache thus getting the latest version.
4) Maybe (3) should be taken into account for designed apps, tests an so on -- transactional request may fail right after restore and the app or test should force a simple request to invalidate the cache.

cc @bmatican @m-iancu to identify next steps.

**Comment by arybochkin** [comment]
I’ve made additional testing related to the time of executing ysqlsh with subsequent “begin transaction + select” statements:

- executed `ysqlsh` before restore and kept open during restore
- executed `ysqlsh` right after the restore had been finished
- executed `ysqlsh` in ~90 seconds after the restore had been finished
- executed `ysqlsh` in ~315 seconds after the restore had been finished
- executed `ysqlsh` in ~615 seconds after the restore had been finished
- executed `ysqlsh` twice right after restore, the first time run `\d` only, second time - the same “begin transaction + select” statements

Every time I got `ERROR: Query error: schema version mismatch for table <uuid>: expected 1, got 0`. But the subsequent quit and then starting `ysqlsh` with “begin transaction + select” statement works fine. So, the cache got updated during the second connection (or on first connection termination?) when the first connection had the error.
As of now it looks like an internal PG cache was not updated for any reason on the first connection after the restore had been finished.

One important thing, `ysqlsh -f YSQLDump` is called during restore procedure to exec DDL, and maybe at this step something got cached without handling subsequent schema updates which are happening in  `CatalogManager::ImportTableEntry()` at [this](https://github.com/yugabyte/yugabyte-db/blob/44a35b99ac0641ad3106a9ba42ee91b30c91101d/ent/src/yb/master/catalog_manager_ent.cc#L1837), [this](https://github.com/yugabyte/yugabyte-db/blob/44a35b99ac0641ad3106a9ba42ee91b30c91101d/ent/src/yb/master/catalog_manager_ent.cc#L1852) and [this](https://github.com/yugabyte/yugabyte-db/blob/44a35b99ac0641ad3106a9ba42ee91b30c91101d/ent/src/yb/master/catalog_manager_ent.cc#L1859) lines.

**Comment by arybochkin** [comment]
postgres log content during the time of restoration
```
I0707 16:41:13.667217 7608 mem_tracker.cc:299] MemTracker: hard memory limit is 9.441395 GB
I0707 16:41:13.667376 7608 mem_tracker.cc:301] MemTracker: soft memory limit is 8.025186 GB
I0707 16:41:13.670583 7608 thread_pool.cc:170] Starting thread pool { name: pggate_ybclient queue_limit: 10000 max_workers: 1024 }
I0707 16:41:13.672983 7608 pggate.cc:149] InitTServerSharedObject: 0, 0, 4
I0707 16:41:13.673609 7608 pg_client.cc:125] Using TServer host_port: 127.0.0.1:9100
I0707 16:41:13.688686 7608 pg_client.cc:132] S 1: Session id acquired
2022-07-07 16:41:18.328 UTC [7608] ERROR: database "yugabyte" already exists
2022-07-07 16:41:18.328 UTC [7608] STATEMENT: CREATE DATABASE yugabyte WITH TEMPLATE = template0 ENCODING = 'UTF8' LC_COLLATE = 'C' LC_CTYPE = 'en_US.UTF-8';
I0707 16:41:18.349052 7623 mem_tracker.cc:299] MemTracker: hard memory limit is 9.441395 GB
I0707 16:41:18.349213 7623 mem_tracker.cc:301] MemTracker: soft memory limit is 8.025186 GB
I0707 16:41:18.352339 7623 thread_pool.cc:170] Starting thread pool { name: pggate_ybclient queue_limit: 10000 max_workers: 1024 }
I0707 16:41:18.354689 7623 pggate.cc:149] InitTServerSharedObject: 0, 0, 4
I0707 16:41:18.355270 7623 pg_client.cc:125] Using TServer host_port: 127.0.0.1:9100
I0707 16:41:18.369145 7623 pg_client.cc:132] S 2: Session id acquired
I0707 16:41:20.623912 7611 poller.cc:66] Poll stopped: Service unavailable (yb/rpc/scheduler.cc:80): Scheduler is shutting down (system error 108)
2022-07-07 16:41:20.646 UTC [7384] LOG: unable to find pgstat entry for abnormally terminated PID 7608
I0707 16:41:22.337070 7623 ybccmds.c:492] Creating Table yugabyte.public.t1
I0707 16:41:22.789955 7626 poller.cc:66] Poll stopped: Service unavailable (yb/rpc/scheduler.cc:80): Scheduler is shutting down (system error 108)
2022-07-07 16:41:22.803 UTC [7384] LOG: unable to find pgstat entry for abnormally terminated PID 7623
```

**Comment by arybochkin** [comment]
postres log for just starting `ysqlsh` with only command `\q`
```
I0707 16:43:02.827679 7906 mem_tracker.cc:299] MemTracker: hard memory limit is 9.441395 GB
I0707 16:43:02.827883 7906 mem_tracker.cc:301] MemTracker: soft memory limit is 8.025186 GB
I0707 16:43:02.831458 7906 thread_pool.cc:170] Starting thread pool { name: pggate_ybclient queue_limit: 10000 max_workers: 1024 }
I0707 16:43:02.834393 7906 pggate.cc:149] InitTServerSharedObject: 0, 0, 4
I0707 16:43:02.835093 7906 pg_client.cc:125] Using TServer host_port: 127.0.0.1:9100
I0707 16:43:02.853317 7906 pg_client.cc:132] S 3: Session id acquired
I0707 16:43:04.264202 7910 poller.cc:66] Poll stopped: Service unavailable (yb/rpc/scheduler.cc:80): Scheduler is shutting down (system error 108)
2022-07-07 16:43:04.276 UTC [7384] LOG: unable to find pgstat entry for abnormally terminated PID 7906
```

**Comment by jasonyb** [comment]
I see two issues:

- **A table schema version mismatch during a transaction does not cause an eventual refresh when that transaction is over.**  This is not the case with (global) catalog version mismatch because, for catalog version, it is checked every query/txn boundary (see `YBCheckSharedCatalogCacheVersion` in the postgres.c loop step 7).  For table schema version mismatch, if it is detected in a transaction, the transaction fails (there is no transaparent retry) and this knowledge of the table schema version mismatch isn't stored anywhere.  So the next transaction that accesses the table will also get table schema version mismatch.  Only when accessing it outside a transaction will the cache refresh happen.

  `YBPrepareCacheRefreshIfNeeded`:

  ```c
      if (YBNeedRetryAfterCacheRefresh(edata))
      {
          ...
          if (consider_retry &&
                  !IsTransactionBlock() &&
                  !YBCGetDisableTransparentCacheRefreshRetry())
          {
  	    ...
              /* Refresh cache now so that the retry uses latest version. */
  ```

  A fix could be to store this knowledge so that we refresh the cache on next query/txn boundary or invalidate the table's cache entry right there because the transaction is aborted anyway.  Also keep in mind the other conditions like `YBNeedRetryAfterCacheRefresh`, `consider_retry`, `YBCGetDisableTransparentCacheRefreshRetry`.  These retry-related conditions should not block cache refreshes for future queries that aren't retries.
- **The table (docdb) schema cache on tserver pg_client_service is not properly invalidated on global cache refresh.**  The table schema is cached in postgres pg_session as `table_cache_` and tserver pg_client_service as `table_cache_` which wraps `cache_`.  A breaking catalog version bump causes a global cache refresh using `YBRefreshCache`, which calls `YBCPgInvalidateCache`, `PgApiImpl::InvalidateCache`, `PgSession::InvalidateAllTablesCache`, `table_cache_.clear();` for the invalidation.  Any future uses of the table will try to open/load it, calling `PgSession::LoadTable` which does

  ```cpp
    auto cached_table_it = table_cache_.find(table_id);
    bool exists = cached_table_it != table_cache_.end();
    if (exists && cached_table_it->second) {
      return cached_table_it->second;
    }
  
    VLOG(4) << "Table cache MISS: " << table_id;
    auto table = VERIFY_RESULT(pg_client_.OpenTable(table_id, exists, invalidate_table_cache_time_));
  ```

  Notice the `exists` variable, which, if you follow down, determines if the tserver pg_client_service will refresh (reopen) its table cache entry.  In case of `PgSession::InvalidateAllTablesCache`, the `table_cache_` entry for the table will not exist, so a refresh will not be requested on tserver's cache.  In case of a table schema version mismatch (that did not happen in a transaction), we call `YBCPgInvalidateTableCacheByTableId`, `PgApiImpl::InvalidateTableCache`, `PgSession::InvalidateTableCache` which requests to refresh (reopen) the tserver's table cache.

  ```cpp
    if (invalidate_on_pg_client) {                                                                          
      // Keep special record about this table_id, so when we would open this table again,                   
      // reopen flag will be sent to pg client service.                                                     
      // This flag means that pg client service should remove table from his cache and fetch                
      // new data from master.                                                                              
      // It is optional optimization, but some tests fails w/o it, since they expect that
      // local table information is updated after alter table operation.                                                                                                                 
      table_cache_[table_id] = nullptr;
  ```

  In short, global catalog refresh does not completely refresh the table schema cache but table schema version mismatch does.

  I believe the tserver's pg_client_service table cache should stay up-to-date.  That was probably the assumption so that we could avoid redundant reopens of the tserver's table cache whenever each postgres backend gets a table schema version mismatch.

**Comment by d-uspenskiy** [comment]
In case of cache refresh postgres invalidates its table cache (and table cache in `PgClientService` as well). Here is the stack trace for such invalidation

```
@     0x7f0457df7a24  yb::pggate::PgSession::InvalidateAllTablesCache() (src/yb/yql/pggate/pg_session.cc:468)
@     0x7f0457d84c72  yb::pggate::PgApiImpl::InvalidateCache() (src/yb/yql/pggate/pggate.cc:491)
@     0x7f0457d6264b  YBCPgInvalidateCache (src/yb/yql/pggate/ybc_pggate.cc:197)
@           0xacd86c  YBRefreshCache (src/postgres/src/backend/tcop/postgres.c:3748)
@           0xace681  YBCheckSharedCatalogCacheVersion (src/postgres/src/backend/tcop/postgres.c:4045)
@           0xac94d6  PostgresMain (src/postgres/src/backend/tcop/postgres.c:5197)
@           0x9fdd4c  BackendRun (src/postgres/src/backend/postmaster/postmaster.c:4560)
@           0x9fce44  BackendStartup (src/postgres/src/backend/postmaster/postmaster.c:4198)
@           0x9fbaaa  ServerLoop (src/postgres/src/backend/postmaster/postmaster.c:1767)
@           0x9f8eb7  PostmasterMain (src/postgres/src/backend/postmaster/postmaster.c:1423)
@           0x904ee7  PostgresServerProcessMain (src/postgres/src/backend/main/main.c:234)
@           0x905311 
@     0x7f04572f8cf2 
@           0x551a6d 
```

As a result current problem can be fixed by changing postgre's sys catalog version after restore procedure.

**Comment by arybochkin** [comment]
Hi @deeps1991, do you have any plans to backport first-level fix to 2.14 or are you going to make all the fixes and then backport alltogether?
